URW-178-API caching

Kenneth-W-Chen opened the pull request:
* Created 2 tables to cache API request responses and a config for each cache
  * So far, only used for Printful API (excludes orders)
* Current PHP functions let you get cached content from the table, add stuff to the cache, clear out the cache (or remove one item from the cache), and add/update cache configs

Kenneth-W-Chen committed:
Create basic logic and table structure

Kenneth-W-Chen committed:
Cleanup table config

Kenneth-W-Chen committed:
Update Printful API to use cache

Kenneth-W-Chen committed:
Add documentation

sebastian-king commented on api/printful/printful.php:
```diff
@@ -8,12 +8,24 @@ public function __construct($printful_api_key = PRINTFUL_API_KEY) {
 
     protected function send_request($URI, $data = false) {
+        $endpoint = 'https://api.printful.com/' . $URI;
+        $isOrder =substr($URI, 0, 6) == 'orders';
```
Let's do a case insensitive regex match here instead, something like `preg_match('/orders$/i, $URI)` or whatever the correct syntax is

sebastian-king commented on api/printful/printful.php:
```diff
@@ -8,12 +8,24 @@ public function __construct($printful_api_key = PRINTFUL_API_KEY) {
+        $isOrder =substr($URI, 0, 6) == 'orders';
+        if(!$isOrder){ // check cache, ignore orders because they shouldn't be cached
+            $r = getCached($endpoint);
+            if($r!== null){ // endpoint is in cache
```
I'd like for the api-cache to be able to handle the cache checking, returning and insertion without needing to implement this everywhere we use the cache. Could the `getCached` method take a instance of `$ch` here where the client just sets the curlopts and it sends the request if needed?

Kenneth-W-Chen replied:
Changed in 7cb4d48. Only caveat is that I made a new object to encapsulate some cURL info and return the object instead of just the content. I only did that because the `PrintfulCustomAPIException`s checked the HTTP response code, which isn't 200 if a cached item is returned (because `curl_exec` wasn't called).
I can have it pass the result only if we switch to throwing generic exceptions instead. Or we could pass the exception type name as a string in the parameter? Let me know if you want one of these instead of the current implementation

sebastian-king commented on api/printful/printful.php:
```diff
@@ -98,6 +112,7 @@ public function get_products($search_string = "") {
         if (!empty($search_string)) {
             $search_string = "&search=" . $search_string;
         }
+        echo 'test';
```
Remove this :)

Kenneth-W-Chen replied:
7cb4d48

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+// these functions use $db, so scripts referencing them need to define it
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint.
```
I love the code comments 👍

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+ * @param string $endpoint The endpoint for the API request
+ * @return array|null The first entry that matches or null if no entry exists
+ */
+function getFromApiCacheTable(string $endpoint)
```
Don't we use snake case everywhere around the site?

sebastian-king replied:
We should probably just name this `getCachedApiResponse` to make it clearer what we're returning

Kenneth-W-Chen replied:
Changed in 7cb4d48 and 15b78fa

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+function getCached(string $endpoint)
+{
+    global $db;
+    $r = getFromApiCacheTable($endpoint);
```
You don't need to do two separate DB queries here, you can just do:
```
SELECT * FROM outgoing_request_cache_config config
JOIN api_cache cache on config.id = cache.config_id
WHERE config.endpoint = ...
```

Kenneth-W-Chen replied:
Changed in 7cb4d48

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+    if ($r) {
+        $id = $r['id'];
+        $queryString = 'UPDATE api_cache 
+                        SET last_successfully_retrieved = UTC_TIMESTAMP, 
```
You might be able to get fancy if you want and set the DB column to auto update the `last_successfully_retrieved` timestamp each time the `content` column updates -- not required at all, just in case you wanted to experiement with that

Kenneth-W-Chen replied:
I can't get `ON UPDATE` to work with `UTC_TIMESTAMP`. Apparently, it's a bug https://bugs.mysql.com/bug.php?id=103228

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+ * Deletes an entry from the cache or deletes all entries from the cache.
+ * @param string|null $endpoint The endpoint of the API request to be removed. If null, empties the cache
+ */
+function clearCached(string $endpoint = null){
```
I'd say leave this commented out if we don't use it, it's a bit risky to have a drop all in the code otherwise

Kenneth-W-Chen replied:
Changed in 7cb4d48

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+    if ($config_name) {
+        return $db->query('INSERT INTO outgoing_request_cache_config (ttl, config_name) VALUES (' . $ttl . ', "' . $db->real_escape_string($config_name) . '")');
+    }
+    return $db->query('INSERT INTO outgoing_request_cache_config (ttl) VALUES (' . $ttl . ')');
```
`config_name` is always required, we should not allow it to be inserted empty, in fact, we should update the SQL to make it a non nullable column

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+ * @param string $endpoint The endpoint for the API request
+ * @return array|false|null The first entry that matches, false if the entry is expired, or null if no entry exists
+ */
+function getCached(string $endpoint)
```
Maybe rename to `getValidCacheEntry` or `getCachedIfNotExpired` or something

Kenneth-W-Chen replied:
Renamed in 7cb4d48

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+        if (gettype($config) == 'integer') {
+            $configId = $config;
+        } else if (gettype($config) == 'string') {
+            $configId = $db->query('SELECT * FROM outgoing_request_cache_config WHERE config_name = "' . $db->real_escape_string($config) . '"')->fetch_array(MYSQLI_ASSOC)['id'];
```
Don't use fetch_array to retreive a single value, I think you want `fetch_field` here which will get just the first value of the first row returned

Kenneth-W-Chen replied:
Removed the query since the function now only accepts the ID

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+                        WHERE id = ' . $id;
+
+    } else {
+        $configId = 1;
```
We should _only_ ever allow insertion by id in this case (due to the way we use the cache, it's always going to be hardcoded, we can always just fetch the id, we want to prioritise consistency and security over developer time), so no need for a sentinal value or dynamic type checking, always require the method is called with a numeric id.

Kenneth-W-Chen replied:
Removed in 7cb4d48

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+            $configId = $db->query('SELECT * FROM outgoing_request_cache_config WHERE config_name = "' . $db->real_escape_string($config) . '"')->fetch_array(MYSQLI_ASSOC)['id'];
+        }
+        $queryString = "INSERT INTO api_cache (endpoint, last_successfully_retrieved, last_attempted_retrieval, config_id, content) 
+                    values ('" . $db->real_escape_string($endpoint) . "', UTC_TIMESTAMP,
```
I'm used to using `CURRENT_TIMESTAMP` for things like this, out of curiosity, is that just an alias for `UTC_TIMESTAMP` or is there actually a timezone difference between those two built in methods?

Kenneth-W-Chen replied:
`CURRENT_TIMESTAMP` returns the timestamp in the server's timezone. I was using UTC since I wasn't sure what the timezone on the server was, and I needed UTC since `time()` in PHP is a Unix timestamp (which is UTC).

sebastian-king commented on api/printful/printful.php:
```diff
@@ -1,5 +1,5 @@
 <?php
-
+require(BASE . '/api/api-cache.php');
```
Is it a guarantee that `BASE` is defined here? I think you might have just been lucky that `BASE` is also provided upstream, and you should actually be importing the template/top first here.
FYI, php will only re-import the template/top here if it wasn't already imported upstream.

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+        }
+        return false;
+    }
+    return null;
```
Oh the beauty of PHP's dynamic and mixed types. I miss it.

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+        $config = $db->query('SELECT * FROM outgoing_request_cache_config WHERE id = ' . $r['config_id'])->fetch_array(MYSQLI_ASSOC);
+        $ttl = $config['ttl'];
+        $now = time();
+        $retrievalTime = strtotime($r['last_successfully_retrieved'] . 'UTC');
```
You've probably already tested this but out of interest, what happens here if/when `last_successfully_retrieved` is null?

Kenneth-W-Chen replied:
It resolves to the current time. I changed it in 7cb4d48

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,132 @@
+        $queryString = "DELETE FROM api_cache";
+    }
+    $q = $db->query($queryString);
+    if ($q === false) {
```
I see that you've sometimes used `$q === false` and sometimes `!$q`, there's a semantic difference between the two that I'm sure you're already aware of (`===` ensures they are the same data type too), we should make sure we are consistent with our usage here.

For a case like checking if the query failed, I'd prefer to use `!$q` because we want to catch all "falsey" cases where $q might be null or otherwise _evaluate_ to false, instead of accidentally assuming a bad query is good, we want to err on the side of caution, basically.

Kenneth-W-Chen replied:
Changed in 7cb4d48

Kenneth-W-Chen committed:
Refactor and make api-cache.php execute cURL instead

Kenneth-W-Chen committed:
Refactor function names

Kenneth-W-Chen committed:
Refactoring. Changed API table's charset

Kenneth-W-Chen commented on sql/outgoing_request_cache.sql:
```diff
@@ -19,4 +19,4 @@ CREATE TABLE `api_cache`(
     PRIMARY KEY (id),
     FOREIGN KEY (config_id) REFERENCES outgoing_request_cache_config(id)
-)AUTO_INCREMENT=8 DEFAULT CHARSET=latin1;
+)AUTO_INCREMENT=8 DEFAULT CHARSET=utf8;
```
Some of the symbols in some Printful API responses are UTF-8. One of the errors lists `\\xE2\\x80\\xB3\\xE2\\x80\\x93...` as an issue. It might actually be a different charset, but given that it's not throwing errors over it anymore, I think UTF-8 will be fine.

sebastian-king replied:
HTTP is always ISO-8859-1 charset, we can transform that into whatever client side charset we feel like. Latin1 should be fine, but if not we can update it to utf-8 if there's a specific problem.

Kenneth-W-Chen replied:
The issue was inserting response contents into the table. MySQL was giving an "incorrect string value" error for some products. The character I mentioned above is the [double prime character](https://www.compart.com/en/unicode/U+2033) which I don't think is present in ISO-8859-1. It's in the product description for every hat product

sebastian-king replied:
Interesting, the curl library must be converting to UTF-8 instead of the standard ISO-8859-1

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,177 @@
+require_once('../template/top.php');
+
+// RegEx to match certain endpoints we want to ignore, e.g., orders
+const IGNORED_ENDPOINTS = '/(?:orders)/i';
```
There's no need to have ignored endpoints, each entry in the db table would be a specific endpoint so if we want to ignore it we just won't make an entry for it.

sebastian-king replied:
I think maybe the confusion here is that you have the code adding a cache entry automatically if its not found, we should not do this. Cache details need to be manually added to the database, if they don't exist and the retrieve cache method is called for that endpoint, we should just throw an error.

If we allow any call to create a cache entry, we can't enforce sensible TTLs, we'll get erroneously created entries, each time an endpoint or API base changes then suddenly we'll get duplicate config entries. We should stay on top of these things manually.

Kenneth-W-Chen replied:
Adjusted in c289fad, specifically, [here (line 73)](https://github.com/untrobotics/website/pull/158/commits/c289fad3a479b93eff024e7c5094ab293f1e3a7b#diff-970b238464763356bddea5f27fef2cba1f6162b327686a4fe165921e8bb448e9R73) and [here (line 87)](https://github.com/untrobotics/website/pull/158/commits/c289fad3a479b93eff024e7c5094ab293f1e3a7b#diff-970b238464763356bddea5f27fef2cba1f6162b327686a4fe165921e8bb448e9R87)

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,177 @@
+    /**
+     * @var bool $curl_executed true if curl_exec was called, false otherwise
+     */
+    public $curl_executed;
```
`curl_executed` is a bad variable name, it could suggest executed is true anytime there isn't an error for example and in the future we may not stay with curl, be as verbose and specific as possible: perhaps, `fetched_new_content` or something

Kenneth-W-Chen replied:
Changed in c289fad. Also changed the respective parameter in the constructor

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,177 @@
+    }
+    $result = curl_exec($ch);
+    // add to cache if no errors and HTTP OK
+    if (!curl_errno($ch) && curl_getinfo($ch, CURLINFO_HTTP_CODE) == 200) {
```
success error codes are anything between 200 and 299, inclusive. for example, 201 and 204 are common success codes alongside 200

Kenneth-W-Chen replied:
c289fad, [here (line 87)](https://github.com/untrobotics/website/pull/158/commits/c289fad3a479b93eff024e7c5094ab293f1e3a7b#diff-970b238464763356bddea5f27fef2cba1f6162b327686a4fe165921e8bb448e9R87)

sebastian-king commented on api/printful/printful.php:
```diff
@@ -1,19 +1,19 @@
 class PrintfulCustomAPI {
     private $api_key;
-
+    private const PRINTFUL_CACHE_CONFIG_ID = 1;
```
Config ID should be unnecessary in the code, we can look everything up using the endpoint URL.

Kenneth-W-Chen replied:
Removed in c289fad. Using the retrieved config ID on [line 82](https://github.com/untrobotics/website/pull/158/commits/c289fad3a479b93eff024e7c5094ab293f1e3a7b#diff-970b238464763356bddea5f27fef2cba1f6162b327686a4fe165921e8bb448e9R82)

sebastian-king commented on api/printful/printful.php:
```diff
@@ -26,23 +26,23 @@ protected function send_request($URI, $data = false) {
         curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
 
-        $result = curl_exec($ch);
+        $cache_result = get_valid_cache_entry($endpoint, $ch, $this::PRINTFUL_CACHE_CONFIG_ID);
```
Aside from the config ID being provided here, this is exactly what I was looking for 👍

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,177 @@
+                                * 
+                            FROM 
+                                api_cache
+                            JOIN 
```
This might need to be a `LEFT JOIN`, a standard `JOIN` I think will return nothing at all if there's no entry in the `api_cache` table but there is an entry in the `outgoing_request_cache_config` table (i.e. JOIN expects both to exist to return anything, but the first time a cache is ever called there won't be an `api_cache` entry just yet)

Kenneth-W-Chen replied:
Changed in c289fad

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,177 @@
+        if ($retrievalTime !== null && $now - strtotime($retrievalTime . 'UTC') < $ttl) {
+            return new CacheResult($r['content']);
+        }
+    }
```
I think if `$q` is false we should probably throw an error

Kenneth-W-Chen replied:
Changed in c289fad [here](https://github.com/untrobotics/website/pull/158/commits/c289fad3a479b93eff024e7c5094ab293f1e3a7b#diff-970b238464763356bddea5f27fef2cba1f6162b327686a4fe165921e8bb448e9R69-R71). Would it be better to return null or treat it as though the endpoint wasn't cached in the first place (i.e., send the cURL request and return that result)?

Kenneth-W-Chen committed:
Fixed for review

sebastian-king commented:
Couple of general notes:

1. I can't actually approve this PR because it's still in draft status
2. You have a mixture of camel case and snake case in your variables names, we use snake case for method and variable names and pascal case for class names here

sebastian-king commented on api/printful/printful.php:
```diff
@@ -1,19 +1,18 @@
-
-        curl_setopt($ch, CURLOPT_URL, 'https://api.printful.com/' . $URI);
+        $endpoint = 'https://api.printful.com/' . $URI;
+        curl_setopt($ch, CURLOPT_URL, $endpoint);
```
It seems tautological to pass `$endpoint` into the `get_valid_cache_entry` method and also set the endpoint before passing this in, we should probably just move this inside of the method.

Kenneth-W-Chen committed:
Refactoring, add endpoint to error_log in insert_cached

sebastian-king commented:
This PR looks good now, can you include the insert SQL for the creation of the configs for printful?

Kenneth-W-Chen commented:
Should I put them in the same file as the table creation or make a separate `cache_configs.sql` file?

Also what would be a good TTL?

sebastian-king commented:
It should go in the `sql/migrations/` directory with a filename like `URW-<ticket num>-cache-configs.sql`
For printful I'd go with a TTL of 24 hours

Kenneth-W-Chen committed:
Add Printful cache config

sebastian-king commented:
Oh, I see now that `endpoint` should be part of the config, not the cache content table. We should have a config entry for each individual endpoint we intend to use.

Kenneth-W-Chen commented:
Are we keeping `endpoint` in the `api_cache` table?
Should we keep `config_name` given that `endpoint` will be similar?

Also, if each config is for a single endpoint, why not put `ttl` in the `api_cache` entry?

Kenneth-W-Chen committed:
Change table columns, add endpoints to config
Changed some SELECTs to grab fewer columns. Added newlines and indents to queries in PHP file to improve readability

Kenneth-W-Chen commented:
Since we're caching Printful stuff now, do we want to remove the `limit` GET param? (It's set to 10 in develop)
https://github.com/untrobotics/website/blob/5d8e2047d11a7be679a755a8c0c4da81f2a8e75b/api/printful/printful.php#L97

sebastian-king commented on api/api-cache.php:
```diff
@@ -0,0 +1,244 @@
+        insert_cached($endpoint, $result, $config_id);
+    }
+
+    return new CacheResult(curl_exec($ch),true, $response_code,curl_errno($ch));}
```
The closing bracket here should be on a new line

Kenneth-W-Chen committed:
Changed tables to separate specific endpoint info from generic info
Rearranged some columns, specified default for columns of type timestamp because of MySQL server settings
Parameterized endpoint strings and added helper function to convert parameterized endpoint strings to the full version
Add varargs for send_request and api-cache functions
Reformatted code

Kenneth-W-Chen committed:
Merge branch 'develop' into URW-178-cache

Kenneth-W-Chen committed:
Change content column type to mediumtext

Kenneth-W-Chen commented:
60c91fe I changed `content`'s type to `mediumtext` because `text` was too small for the Gear page, Specifically, the "Aerospace Division T" is over 200kB. No other product has this issue except for that one. 

The specific cause is because the product has 28 variants, and Printful returns info for each variant for the single request. The data is repetitive (e.g., each variant lists if it's available in the US, UK, etc.). I can compress it into a normal zip down to 10kB or into a `.7z` to about 8kB.

Kenneth-W-Chen committed:
Compress content column and change type to blob

Kenneth-W-Chen committed:
refactor/reformating

sebastian-king committed:
URW-178: fix naming and serialisation

sebastian-king committed:
URW-64: fix naming conflicts in sql migration

sebastian-king committed:
Resolve merge conflicts
